From c1cac21cfdb52c17b663a4daf266298f80cae6fb Mon Sep 17 00:00:00 2001
Message-Id: <c1cac21cfdb52c17b663a4daf266298f80cae6fb.1709043678.git.dpark@linux.microsoft.com>
From: Dongsu Park <dpark@linux.microsoft.com>
Date: Tue, 27 Feb 2024 15:16:26 +0100
Subject: [PATCH 01/13] Revert "x86/efistub: Use 1:1 file:memory mapping for
 PE/COFF .compat section"

This reverts commit 0a962f2fbaa976af9eed21d0306370cded485787.
---
 arch/x86/boot/header.S | 14 ++++++++------
 arch/x86/boot/setup.ld |  6 +++---
 2 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/arch/x86/boot/header.S b/arch/x86/boot/header.S
index a1bbedd9..b2771710 100644
--- a/arch/x86/boot/header.S
+++ b/arch/x86/boot/header.S
@@ -106,7 +106,8 @@ SYSSEG		= 0x1000		/* historical load address >> 4 */
 	.word	0				# MinorSubsystemVersion
 	.long	0				# Win32VersionValue
 
-	.long	setup_size + ZO__end		# SizeOfImage
+	.long	setup_size + ZO__end + pecompat_vsize
+						# SizeOfImage
 
 	.long	salign				# SizeOfHeaders
 	.long	0				# CheckSum
@@ -142,7 +143,7 @@ SYSSEG		= 0x1000		/* historical load address >> 4 */
 	.ascii	".setup"
 	.byte	0
 	.byte	0
-	.long	pecompat_fstart - salign 	# VirtualSize
+	.long	setup_size - salign 		# VirtualSize
 	.long	salign				# VirtualAddress
 	.long	pecompat_fstart - salign	# SizeOfRawData
 	.long	salign				# PointerToRawData
@@ -155,8 +156,8 @@ SYSSEG		= 0x1000		/* historical load address >> 4 */
 #ifdef CONFIG_EFI_MIXED
 	.asciz	".compat"
 
-	.long	pecompat_fsize			# VirtualSize
-	.long	pecompat_fstart			# VirtualAddress
+	.long	8				# VirtualSize
+	.long	setup_size + ZO__end		# VirtualAddress
 	.long	pecompat_fsize			# SizeOfRawData
 	.long	pecompat_fstart			# PointerToRawData
 
@@ -171,16 +172,17 @@ SYSSEG		= 0x1000		/* historical load address >> 4 */
 	 * modes this image supports.
 	 */
 	.pushsection ".pecompat", "a", @progbits
-	.balign	salign
+	.balign	falign
+	.set	pecompat_vsize, salign
 	.globl	pecompat_fstart
 pecompat_fstart:
 	.byte	0x1				# Version
 	.byte	8				# Size
 	.word	IMAGE_FILE_MACHINE_I386		# PE machine type
 	.long	setup_size + ZO_efi32_pe_entry	# Entrypoint
-	.byte	0x0				# Sentinel
 	.popsection
 #else
+	.set	pecompat_vsize, 0
 	.set	pecompat_fstart, setup_size
 #endif
 	.ascii	".text"
diff --git a/arch/x86/boot/setup.ld b/arch/x86/boot/setup.ld
index 3a2d1360..83bb7efa 100644
--- a/arch/x86/boot/setup.ld
+++ b/arch/x86/boot/setup.ld
@@ -24,9 +24,6 @@ SECTIONS
 	.text		: { *(.text .text.*) }
 	.text32		: { *(.text32) }
 
-	.pecompat	: { *(.pecompat) }
-	PROVIDE(pecompat_fsize = setup_size - pecompat_fstart);
-
 	. = ALIGN(16);
 	.rodata		: { *(.rodata*) }
 
@@ -39,6 +36,9 @@ SECTIONS
 	. = ALIGN(16);
 	.data		: { *(.data*) }
 
+	.pecompat	: { *(.pecompat) }
+	PROVIDE(pecompat_fsize = setup_size - pecompat_fstart);
+
 	.signature	: {
 		setup_sig = .;
 		LONG(0x5a5aaa55)
-- 
2.39.2

